connection-loop.ts

```diff
@@ -0,0 +1,262 @@
+        try {
+          await invokePush();
+          lastConnectionErrored = false;
+        } catch {
+          lastConnectionErrored = true;
+        }
+        const stop = Date.now();
+        delay = computeDelayAndUpdateDurations(
+          delay,
+          start,
+          stop,
+          !lastConnectionErrored,
+          durations,
+        );
+        console.log(
+          'Last connection took',
+          stop - start,
+          'ms. New delay is ',
+          delay,
+        );
+        pushCounter--;
+        wakeUpWaitingPushes();
+        this._currentPushResolver.resolve();
+      })();
+    }
+  }
+}
+
+class CancelToken {
+  readonly promise: Promise<void>;
+  readonly cancel: () => void;
+  constructor() {
+    ({promise: this.promise, resolve: this.cancel} = resolver());
+  }
+}
+
+function sleep(ms: number, cancelPromise?: Promise<unknown>) {
+  let id: ReturnType<typeof setTimeout> | 0 = 0;
+  let reject: (message: Error) => void;
+  const promise = new Promise<void>((resolve, rej) => {
+    reject = rej;
+    id = setTimeout(() => {
+      id = 0;
+      resolve();
+    }, ms);
+  });
+
+  if (cancelPromise) {
+    cancelPromise.then(() => {
+      if (id) {
+        clearTimeout(id);
+        reject(new Error('sleep cancelled'));
+      }
+    });
+  }
+  return promise;
+}
+
+let shouldError = 0;
+
+async function invokePush() {
+  console.log('push ->', new Date());
+  await sleep(100 + Math.random() * 100);
+  if (Math.random() < shouldError) {
+    await sleep(1000);
+    console.log('<- push ERROR', new Date());
+    throw new Error('Intentional error');
+  }
+  console.log('<- push', new Date());
+}
+
+const waitingConnectionResolvers: Set<() => void> = new Set();
+
+function wakeUpWaitingPushes() {
+  if (waitingConnectionResolvers.size > 0) {
+    console.log('wakeUpWaitingPushes', waitingConnectionResolvers.size);
+  }
+  for (const resolve of waitingConnectionResolvers) {
+    waitingConnectionResolvers.delete(resolve);
+    resolve();
+  }
+}
+
+function waitUntilAvailableConnections() {
+  const {promise, resolve} = resolver();
+  waitingConnectionResolvers.add(resolve);
+  return promise;
+}
+
+export function resolver<R = void>(): {
+  promise: Promise<R>;
+  resolve: (res: R) => void;
+} {
+  let resolve!: (res: R) => void;
+  const promise = new Promise<R>(res => {
+    resolve = res;
+  });
+  return {promise, resolve};
+}
+
+function computeDelayAndUpdateDurations(
+  delay: number,
+  start: number,
+  stop: number,
+  ok: boolean,
+  durations: Durations,
+): number {
+  const last: Duration | undefined = durations[durations.length - 1];
+  durations.push({start, stop, ok});
+  if (durations.length === 1) {
+    return (stop - start) / MAX_CONNECTIONS;
+  }
+  if (!ok) {
+    return Math.min(MAX_DELAY, delay * 2);
+  }
+
+  if (!last.ok) {
+    // Recovered
+    return MIN_DELAY;
+  }
+
+  const okDurations = durations.filter(({ok}) => ok);
+  const average =
```

Comment:
Seems like median would be better.

Reply:
Sure. Also, we talked about maybe using data over the last n seconds etc.